cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -654,7 +655,8 @@ TEST(Substrait, ReadRel) {
   ASSERT_EQ(scan_node_options.dataset->type_name(), "filesystem");
   const auto& dataset =
       checked_cast<const dataset::FileSystemDataset&>(*scan_node_options.dataset);
-  EXPECT_THAT(dataset.files(), ElementsAre("/tmp/dat1.parquet", "/tmp/dat2.parquet"));
+  EXPECT_THAT(dataset.files(),
+              UnorderedElementsAre("/tmp/dat1.parquet", "/tmp/dat2.parquet"));
   EXPECT_EQ(dataset.format()->type_name(), "parquet");
   EXPECT_EQ(*dataset.schema(), Schema({field("i", int64()), field("b", boolean())}));
 }
```

Comment:
I'm surprised: we aren't adding any tests for globbing here?

Reply:
To add glob tests here, we should create directories and files to perform glob discovery. Although I'm not sure if that's what we want to do . CC @westonpace

Reply:
Well, otherwise it means the functionality is untested?

Reply:
I think is only tested for file type because they don't need discovery. For directories and globs is not tested. I think directories tests are cover by each `filesystem_test` with `GetFileInfo(Selector selector)`. I think we can add a similar way of testing this, something like:
1. Create filesystem
2. Add directories and files
3. Check matches between patters and paths

Reply:
Another option is to do the tests indirectly in filesystem_test.

Reply:
The filesystem tests don't test the globbing logic you've implemented on the substrait side.

Reply:
`GetGlobFiles` can be called from filesystem tests, in specific, local filesystem tests, and it'd look something like this:
```
TYPED_TEST(TestLocalFS, GetGlobFiles) {
  ASSERT_OK(this->fs_->CreateDir("AB/CD"));
  CreateFile(this->fs_.get(), "AB/CD/a.txt", "data");
  CreateFile(this->fs_.get(), "AB/CD/ab.txt", "data");
  CreateFile(this->fs_.get(), "AB/CD/abc.txt", "data");

  std::vector<FileInfo> infos;
  ASSERT_OK_AND_ASSIGN(infos, GetGlobFiles(this->fs_.get(), "AB/CD/*b*.txt"));
  ASSERT_EQ(infos.size(), 2);
  AssertFileInfo(infos[0], "AB/CD/ab.txt", FileType::File, 4);
  AssertFileInfo(infos[1], "AB/CD/abc.txt", FileType::File, 4);
}
```

Reply:
Ok, then I think `GetGlobFiles` should be moved into filesystem utilities first.

Reply:
I agree that the majority of tests should go into the filesystems utilities tests.  It would be nice if we could have at least one test in the serde logic just to make sure we are handling the case where a glob URI is specified (e.g. coverage for that specific switch case branch, parsing the URI to extract the glob pattern) but I recognize that gets more complicated.

It might make more sense to add these tests in when we add non-local filesystem support to Substrait.  I just created ARROW-16463.  Can you add a note to that JIRA about making sure we test the different types of file options?

Reply:
Sure